### PR TITLE
feat(bindings): Exposing Register Notification Handler in UniFFI

### PR DESCRIPTION
**Pull request by Velin92**
This allows to listen for the ClientDelegate to listen for push notifications.
The function should be called in a concurrent context since it blocks execution.
For the handler to work properly the Client will need to add m.room.member (at least with the value $ME) and m.room.power_levels events to the Sliding Sync Required States

Also exposed all the events, but did not expose a content for all of them for now, just for the ones I might need for a basic implementation of the simplest notification feature of El-X

### CODE REVIEW

File: bindings/matrix-sdk-ffi/src/event.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
+use anyhow::{bail, Context};
+use ruma::events::{AnySyncMessageLikeEvent, AnySyncTimelineEvent};
+
+use crate::{ClientError, MessageType};
+
+pub struct TimelineEvent(pub(crate) AnySyncTimelineEvent);
+
+#[uniffi::export]
```
Comment (Velin92) [inline bindings/matrix-sdk-ffi/src/event.rs:8]: @jplatte Why the attribute here gives me an error?
It says:

required for `std::sync::Arc<event::TimelineEvent>` to implement `uniffi::FfiConverter<UniFfiTag>`

However I have defined in the .dul an empty interface implementation just as for RoomMember or TimelineItem

Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/event.rs:8]: The empty interface definition inside the UDL does the right thing, but only if `TimelineEvent` is visible in the place the UDL scaffolding is included (the crate root). You should add `event::*` to the import / re-export at line 54 of `lib.rs`.

File: bindings/matrix-sdk-ffi/src/event.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                let state_key = content.state_key().to_string();
+                let original_content = content
+                    .as_original()
+                    .context("Failed to get original content")?
+                    .content
+                    .clone();
+                StateEventContent::RoomMember {
+                    user_id: state_key,
+                    membership_state: original_content.membership.into(),
+                }
```
Comment (Velin92) [inline bindings/matrix-sdk-ffi/src/event.rs:89]: @jplatte 
Had to implement this quite a lot, but I am unable to implement a generic function that does this automatically, I get this error on this implementation:
```
fn try_convert<T: StaticStateEventContent + RedactContent>(
    event: ruma::events::SyncStateEvent<T>,
) -> anyhow::Result<T>
where
    T::Redacted: RedactedStateEventContent,
{
    let original_content =
        event.as_original().context("Failed to get original content")?.content.clone();
    Ok(original_content)
}
```

the method `as_original` exists for enum `SyncStateEvent<T>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`<<T as ruma::events::RedactContent>::Redacted as ruma::events::RedactedStateEventContent>::StateKey = <T as ruma::events::StateEventContent>::StateKey`

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 3]
```diff
@@ -547,6 +548,27 @@ impl Client {
...
         })
     }
+
+    /// Registers a notification handler on the delegate if required
+    /// the sliding sync requires to have registered m.room.member with value
+    /// $ME and m.room.power_levels to be able to intercept the events.
+    /// This function blocks execution and should be dispatched concurrently
+    pub fn register_notification_handler(&self) {
+        let delegate = Arc::clone(&self.delegate);
+        let handler = move |notification, room: SdkRoom, _| {
```
[the remark this answers is not in the document]
Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/client.rs:558]: Shouldn't there be a sliding sync view that gets you all of the events for "active" (top of the room list) rooms? And rooms where you get notified should be active by definition, no?

File: bindings/matrix-sdk-ffi/src/api.udl [hunk 4]
```diff
@@ -7,8 +7,23 @@ interface ClientError {
 
 callback interface ClientDelegate {
     void did_receive_auth_error(boolean is_soft_logout);
+    void did_receive_notification(NotificationItem notification);
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/api.udl:10]: Why is this part of the same delegate?

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 5]
```diff
@@ -547,6 +548,27 @@ impl Client {
             Ok(user_profile)
         })
     }
+
+    /// Registers a notification handler on the delegate if required
+    /// the sliding sync requires to have registered m.room.member with value
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client.rs:553]: Are these supposed to be separate sentences?

```suggestion
    /// Registers a notification handler on the delegate if required.
    ///
    /// The sliding sync requires to have registered m.room.member with value
```

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 6]
```diff
@@ -547,6 +548,27 @@ impl Client {
...
+    /// $ME and m.room.power_levels to be able to intercept the events.
+    /// This function blocks execution and should be dispatched concurrently
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/client.rs:555]:
```suggestion
    /// This function blocks execution and should be dispatched concurrently.
```

File: bindings/matrix-sdk-ffi/src/event.rs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        };
+        Ok(event_type)
+    }
+}
+
+#[derive(uniffi::Enum)]
+pub enum TimelineEventType {
+    MessageLike { content: MessageLikeEventContent },
+    // we don't support state events yet so they do not have any content
+    State { content: StateEventContent },
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/event.rs:39]: But it says `content: StateEventContent` right there :D

File: bindings/matrix-sdk-ffi/src/notification_service.rs [hunk 8]
```diff
@@ -35,7 +67,6 @@ impl NotificationService {
     }
 }
 
-#[uniffi::export]
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/notification_service.rs:38]: Why did you remove this?

File: bindings/matrix-sdk-ffi/src/timeline.rs [hunk 9]
```diff
@@ -449,82 +450,97 @@ pub struct Message(matrix_sdk::room::timeline::Message);
 #[uniffi::export]
 impl Message {
     pub fn msgtype(&self) -> Option<MessageType> {
+        if let Ok(message_type) = self.0.msgtype().clone().try_into() {
+            Some(message_type)
+        } else {
+            None
+        }
```
Comment (jplatte) [inline bindings/matrix-sdk-ffi/src/timeline.rs:457]:
```suggestion
        self.0.msgtype().clone().try_into().ok()
```

File: bindings/matrix-sdk-ffi/src/notification_service.rs (hunk 8 above)
Comment (Velin92) [reply bindings/matrix-sdk-ffi/src/notification_service.rs:38]: putting it back

File: bindings/matrix-sdk-ffi/src/api.udl (hunk 4 above)
Comment (Velin92) [reply bindings/matrix-sdk-ffi/src/api.udl:10]: shoul it be a different one?
It just made the implementation a bit easier to use the same existing one

File: bindings/matrix-sdk-ffi/src/event.rs (hunk 7 above)
Comment (Velin92) [reply bindings/matrix-sdk-ffi/src/event.rs:39]: let's remove this old comment

File: bindings/matrix-sdk-ffi/src/api.udl (hunk 4 above)
Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/api.udl:10]: To me it makes more sense to have different ones, so they're basically just replacements for closures. A client doesn't necessarily always want to receive both.

File: bindings/matrix-sdk-ffi/src/client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/client.rs:559]: Hm, maybe we can discuss in the Sync later. I'm not sure how hard it will be to take the room name from the sliding sync list.

[the remark this answers is not in the document]
Comment (Velin92) [reply bindings/matrix-sdk-ffi/src/client.rs:568]: Otherwise... wouldnt't be possible for the SS handling to just update the SdkRoom content directly? (ofc I imagine this is something that would need to be addressed into a separate PR)

[the remark this answers is not in the document]
Comment (jplatte) [reply bindings/matrix-sdk-ffi/src/client.rs:568]: Not easily. We store all room metadata that comes from events as (simplified) events. This is important for instance so we know when a redaction comes in, we can check if it's about the room name, and update the room data accordingly.

**Review by jplatte — APPROVED** [review]
